### ISSUE TITLE
Ability to set API/Search requests to faster than 1 minute between each

### ISSUE THREAD

**Issue by Marthisdil** [mylar3/mylar3#441]
It would be great if we could have e the ability to have the search function search faster that once per minute, on a per-search-engine basis.

For example, NZBGeek paid subscriptions don't have an API limit really...so it'd be awesome to be able to increase the number of searches per minute (or reduce the time per search to one per second, or 5 seconds or something.

**Comment by BRTPOB** [comment]
You can do that if you change the indexer to include [local] in the name.

**Comment by evilhero** [comment]
It's not so much about the number of API requests. If you have a paid indexer, those are usually unlimited (unless ya know, they're crappy).

Almost every indexer does however limit how often you make a request against their API (rate-limiting). This is described in their TOS, which I know most people don't read, but it does exist so that they can quote it when an account gets disabled.

That being said Mylar can only search for one item at a time, which is just the searching aspect, not downloading or post-processing - however, each search is rate limited to 60s on purpose. 

The best that could be done is to submit a search request to each indexer almost simultaneously, and then aggregate the results - which was in place however, that was in regards to manual selection, not automatic selection. There was still a delay in compiling the aggregated results tho (30s I believe). 

But in either method, because Mylar has to search by string query, it has to run thru all iterations of an issue #, and possible alternative search names (as 001 does not equal 01,which does not equal 1). So searching for just 1 issue could bring anywhere from 1-10 separate API hits depending on the # and naming aspects. 

Do a thousand searches (a typical wanted list for some) like that, with no limiter, you'll end up getting banned by the indexer usually - or if it happens enough the client will be outright banned. 

I think adding the [local] tag to a newznab entry should work with domain addresses as @BRTPOB mentioned, however that would bypass the rate limiting per query, which would be at your own risk.

**Comment by Marthisdil** [comment]
Evilhero: thanks for the reply. "It's not so much about the number of API requests. If you have a paid indexer, those are usually unlimited (unless ya know, they're crappy).

Almost every indexer does however limit how often you make a request against their API (rate-limiting). This is described in their TOS, which I know most people don't read, but it does exist so that they can quote it when an account gets disabled.

That being said Mylar can only search for one item at a time, which is just the searching aspect, not downloading or post-processing - however, each search is rate limited to 60s on purpose."

I do understand rate limits, I also understand searching on variants in issue numbers, etc.  I've found all "episode" typoe of processors have to do similar.

However, with good indexers, API rate limits are fairly generous.  For example, with paid membership to NZBGeek, here are the limits: 

"The limits are like this:

More than 2 API calls per SECOND: this will result in a temporary suspension of your API for 60 seconds.

More than 250 API calls IN 5 MINUTES: this will result in the suspension of your API access."

I know others have different numbers (thus why my comment was on a per-search engine basis).

I guess why not take advantage of the ability to parallelize the searching?  if users can set their search limits based on the search engines, it would greatly increase the speed of searching, while staying within limits dictated by the indexers.

Just my 2 cents.  Thanks for the great software :)

**Comment by evilhero** [comment]
Right now tagging the indexer as [local] would bypass the speed restriction regardless of any other setting on a per-indexer basis. Building in an optional way to set the search delay per indexer is a bit more of a process unfortunately due to how the program stores all the items in the config.ini. It can be done, don't get me wrong - just that there's quite a few things that need to get modified / changed in order to have the option available (and in the GUI).

Previously, years ago I guess at this point, the per SECOND aspect was definitely higher than that - which is why we defaulted to 60s spacings (as an example, althub says ```To avoid being banned please setup your search tools to ping the API once every 10 minutes, at a minimum.``` but then in another section they say you can hit the API as often as you want - so definitely conflicting stories). Plus at the time headphones was getting a bad rap for hammering providers without any restrictions and was being banned by the providers - so we were trying to distance ourselves from that program by opting for a more favourable approach with Mylar. 

Plus it's comics - I mean waiting an extra minute isn't the end of the world since it was all automated anyways with no user interaction on the search front (I had something in there at one point that involved manual selection of all the results when searching, but there were display problems happening after the results were displayed so it was temporarily scrapped - but will come back in v3 at some point since the code is still in there).

So yeah , we can look at adding this option in - but there are some other items that we have to finish first prior to going down this road unfortunately.

**Comment by DieselTech** [comment]
It's not a weird thing to have sane rate limits. These limits exist because indexs would ban mylar if they didn't. And just because your indexer is unlimited in terms of downloads does not mean they allow hammering their API with requests. Velocity is different than consumption. Mylar is meant to be a "in the background" type of program that just sits and searches over time. If you want instant results then download the file manually and post-process it.

**Comment by falo2k** [comment]
Let alone the issue that if you're doing any metadata lookups on ComicVine then you're going to be squeezed far tighter than any  indexer limits because of their **very** restrictive API limits.

**Comment by tomerh2001** [comment]
> @tomerh2001 This is exactly the reason why I won't willy-nilly change it to an open value because people will abuse it and set it for some insanely low amount (0.1s, really?) not realizing or caring how an api actually works with load rates. I don't know you or care about you to go to bat for you when the indexers email me to complain about the app hitting their api too quickly from multiple users - I've done that in the past, and I'm not doing it again. 
> 
> If you didn't read the reponse above, text based querying is far, far different than ID-based queries. As such they allow more frequent hits than text based as text based queries will return a wide array of results and can be taxing on their api. TV/movies/music all allow ID based queries on newznab providers. 
> Comics have never had the ability to query by ID. 
> 
> If you have multiple indexers set up in mylar, your better off to only have a couple since the way mylar works is it goes thru each indexer in sequence doing several requests per issue per indexer. This can take one issue up to 10mins or more to search because of the text based searching (1,01,001, altnames, overlapping years). Don't use torrents (cause they suck for comics), and I'm betting you have a bunch of indexers in mylar that don't even have comics on them, all of which contribute to longer search times. If the indexer that has comics is listed fifth in your provider order, it has to query the first 4 a bunch of times before even getting to it. 

How does me, sending queries to my indexers, results in you getting an email exactly?

If I want to try things out and eventually get myself blocked from my indexers, why should you care about it?

Also, its open source, I can just remove limit myself and run that 🤷‍♂️

Unless I'm missing something in how exactly you are related to what I chose to do with mylar, this is a very weird chat

**Comment by DieselTech** [comment]
> How does me, sending queries to my indexers, results in you getting an email exactly?

Because the program making the request identifies itself as mylar. When the indexers want to see what it misbehaving, they look at the logs and find the bad actor.

**Comment by evilhero** [comment]
You're using the app that has the github repo in it as well as a user agent.. Obviously you're not that dense that you can't figure out that indexers can view the user agents being sent with each request, thousands of requests per minute all of comics is going to lead them to look at the requests more in detail. They will talk to the app dev to get confirmation of why it's happening if it's brought up by users as to why they're banned. 

Your casual ignorance of the way that mylar works pretty much shows you have no clue as to how it actually works.

It is open source, you do you. I gave you some options to use to help, but you can do w/e. 

Why even bother to ask the question since you don't care about the response?

**Comment by BRTPOB** [comment]
Additionally, @tomerh2001, you could have easily used the (not great) workaround I provided 3.5 years ago.

**Comment by barbequesauce** [comment]
Closing due to staleness, we welcome contributions for new features. If this is an issue and you are still experiencing it, please open a new ticket with a care package attached.
